rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,432 @@
+		newBytesRead, err := IOMgrConn.Read(prefixBuf[bytesRead : bytesRead+1])
+		if newBytesRead == 0 {
+			if io.EOF == err {
+				return bytesRead, n, nil
+			} else if err != nil {
+				return bytesRead, n, err
+			}
+			// A Reader should not return (0, nil), but if it does,
+			// it should be treated as no-op (according to the
+			// Reader contract). So let's go on...
+			continue
+		}
+		bytesRead += newBytesRead
+		// Now present everything read so far to the varint decoder and
+		// see if a varint can be decoded already.
+		n, varIntBytes = proto.DecodeVarint(prefixBuf[:bytesRead])
+	}
+
+	return bytesRead, n, err
+}
+
+func (IOMgrConn *IOMgrConnection) writeMessage(messageTypeID uint64, m proto.Message) (int, error) {
+
+	// Function writeMessage()
+	// source: https://github.com/stashed/stash/blob/master/vendor/github.com/matttproud/golang_protobuf_extensions/pbutil/encode.go
+	//
+	// Copyright 2013 Matt T. Proud
+	//
+	// Licensed under the Apache License, Version 2.0 (the "License");
+	// you may not use this file except in compliance with the License.
+	// You may obtain a copy of the License at
+	//
+	//     http://www.apache.org/licenses/LICENSE-2.0
+	//
+	// Unless required by applicable law or agreed to in writing, software
+	// distributed under the License is distributed on an "AS IS" BASIS,
+	// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	// See the License for the specific language governing permissions and
+	// limitations under the License.
+
+	var buf [binary.MaxVarintLen32]byte
+	encodedLength := binary.PutUvarint(buf[:], messageTypeID)
+	IOMgrConn.Write(buf[:encodedLength])
+
+	em, err := proto.Marshal(m)
+	if nil != err {
+		return 0, err
+	}
+
+	encodedLength = binary.PutUvarint(buf[:], uint64(proto.Size(m)))
+
+	vib, err := IOMgrConn.Write(buf[:encodedLength])
+	if nil != err {
+		return vib, err
+	}
+
+	pbmmb, err := IOMgrConn.Write(em)
+
+	return vib + pbmmb, err
+}
+
+func (IOMgrConn *IOMgrConnection) readMessage(messageTypeID uint64, m proto.Message) (n int, err error) {
+
+	// Function readMessage()
+	// source: https://github.com/stashed/stash/blob/master/vendor/github.com/matttproud/golang_protobuf_extensions/pbutil/decode.go
+	//
+	// Copyright 2013 Matt T. Proud
+	//
+	// Licensed under the Apache License, Version 2.0 (the "License");
+	// you may not use this file except in compliance with the License.
+	// You may obtain a copy of the License at
+	//
+	//     http://www.apache.org/licenses/LICENSE-2.0
+	//
+	// Unless required by applicable law or agreed to in writing, software
+	// distributed under the License is distributed on an "AS IS" BASIS,
+	// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	// See the License for the specific language governing permissions and
+	// limitations under the License.
+
+	var prefixBuf [binary.MaxVarintLen64]byte
+	var bytesRead, varIntBytes int
+	var messageLength uint64
+	var readMessageTypeID uint64
```

Comment:
32-bit

Reply:
@cxxman the thing is that the method DecodeVarint, decode both: varint32 and varint64 and return a varint64: https://pkg.go.dev/github.com/golang/protobuf/proto#DecodeVarint
There is no function only to decode a varint32, so we would have to develop it ourselves. Does it worth it?

Reply:
It doesn't do it correctly for varint32. I already explained in some other comment - you need to add parameter, to which extent to read. it will be either `binary.MaxVarintLen32` or `binary.MaxVarintLen64`.